db/etcd.go

```diff
@@ -569,259 +406,177 @@ func (etcddb etcdDB) InitTestRun(testUuid string, testAgentMap map[string]map[st
-
-	_, err = kapi.Set(
+func (etcddb *etcdDB) SetAgentTestData(testUUID, agentUUID, testData string) error {
+	_, err := etcddb.keysAPI.Set(
 		context.Background(),                                                     // context
-		fmt.Sprintf("/todd/testruns/%s/agents/%s/testdata", testUuid, agentUuid), // key
+		fmt.Sprintf("/todd/testruns/%s/agents/%s/testdata", testUUID, agentUUID), // key
 		testData, // value
 		nil,      //optional args
 	)
 	if err != nil {
-		log.Errorf("Problem updating testdata for agent %s in test %s", agentUuid, testUuid)
+		log.Errorf("Problem updating testdata for agent %s in test %s", agentUUID, testUUID)
 		log.Error(err)
-		os.Exit(1)
+		return err
 	}
 
 	return nil
 }
 
 // GetTestStatus returns a map containing a list of agent UUIDs that are participating in the provided test, and their status in this test.
-func (etcddb etcdDB) GetTestStatus(testUuid string) map[string]string {
+func (etcddb *etcdDB) GetTestStatus(testUUID string) (map[string]string, error) {
 
-	ret_map := make(map[string]string)
+	retMap := make(map[string]string)
 
-	etcd_loc := fmt.Sprintf("http://%s:%s", etcddb.config.DB.IP, etcddb.config.DB.Port)
-	cfg := client.Config{
-		Endpoints: []string{etcd_loc},
-		Transport: client.DefaultTransport,
-		// set timeout per request to fail fast when the target endpoint is unavailable
-		HeaderTimeoutPerRequest: time.Second,
-	}
-	c, err := client.New(cfg)
-	if err != nil {
-		log.Fatal(err)
-	}
-	kapi := client.NewKeysAPI(c)
+	keyStr := fmt.Sprintf("/todd/testruns/%s/agents", testUUID)
 
-	key_str := fmt.Sprintf("/todd/testruns/%s/agents", testUuid)
+	log.Debug("Retrieving detailed test status for ", testUUID)
 
-	log.Debug("Retrieving detailed test status for ", testUuid)
-
-	resp, err := kapi.Get(context.Background(), key_str, &client.GetOptions{Recursive: true})
+	resp, err := etcddb.keysAPI.Get(context.Background(), keyStr, &client.GetOptions{Recursive: true})
 	if err != nil {
 		fmt.Println(err)
-		log.Errorf("Error - empty test encountered: %s", testUuid)
-		return ret_map
-	} else {
-
-		log.Debugf("Etcd 'get' is done. Metadata is %q\n", resp)
-
-		// We are expecting that this node is a directory
-		if resp.Node.Dir {
-
-			// Iterate over found objects
-			for i := range resp.Node.Nodes {
-				thisAgent := resp.Node.Nodes[i]
-
-				statusKey := fmt.Sprintf("%s/status", thisAgent.Key)
-
-				// Extract UUID from key string
-				agentUuid := strings.Replace(thisAgent.Key, fmt.Sprintf("/todd/testruns/%s/agents/", testUuid), "", 1)
+		log.Errorf("Error - empty test encountered: %s", testUUID)
+		return retMap, nil
+	}
 
-				statusResp, err := kapi.Get(context.Background(), statusKey, nil)
-				if err != nil {
-					log.Errorf("Error - empty agent status encountered: %s", testUuid)
-				}
+	log.Debugf("Etcd 'get' is done. Metadata is %q\n", resp)
 
-				ret_map[agentUuid] = statusResp.Node.Value
+	// We are expecting that this node is a directory
+	if !resp.Node.Dir {
```

Comment:
NOTE - go back and make sure these are all negated (just me being obsessive, I am 99 percent sure they are)